### ISSUE TITLE
Training setup too slow, makes it painful when we have to debug

### ISSUE THREAD

**Issue by ichn-hu** [allenai/allennlp#2546]
I am currently using allennlp to develop my own model based on these well written and documented models in the model zoo, I really like AllenNLP and enjoy the convenience it brings. However it is very painful to set everything work at the beginning, because it is so slow (roughly 3 to 5 mins) to setup things such as load a pretrained model, read the dataset, load a pretrained embedding file etc to come to the point where things does not work and I need to figure it out, and whenever I changed something (most of them in forward functions) it takes such a long period to see if it works or not, this is so painful to me! 

**Describe the solution you'd like**

A clever solution for this might be caching the setup to memory, such as /dev/shm, once the setup is ready, it typically won't change for debugging training routine, so we can happily cache them and don't have to rerun them at each subsequent run.

**Comment by joelgrus** [comment]
In our own work we typically develop against a very small dataset (e.g, a handful of examples) with very small embeddings (e.g. with a tiny vocabulary), which allows you to iterate very quickly to get to a "working" model, check that loss is decreasing, and so on. Only once that's all working do we switch to the full dataset. I can't speak for everyone here, but my sense is that this workflow is pretty effective and relatively pain-free.

I think caching a general "setup" is a hard problem. It's possible we might push in that direction at some point, but it's unlikely to be a priority in the near term.

**Comment by matt-gardner** [comment]
I will just second what @joelgrus said, and point out that the mechanism we use to do this is our tests.  We make this very easy with our test infrastructure - the first thing that I do when writing a new model is set up a test, that's typically about 10 lines of code: https://github.com/allenai/allennlp/blob/540ebac72c430aebcc595bd9909306de0181556e/allennlp/tests/models/atis_semantic_parser_test.py#L7-L15

There's also some setup in constructing the test fixtures, but these are just very small versions of the data and configuration that you'll already be using.

**Comment by scarecrow1123** [comment]
I guess the OP is asking for something like `make-vocab`. @ichn-hu Have you tried it? It has come in handy for me a lot of times. This essentially does the vocab creation and could save some time when you have large datasets. It creates a `vocabulary` directory which you can specify in the configuration when you start training. Details [here](https://github.com/allenai/allennlp/blob/580dc8b0e2c6491d4d75b54c3b15b34b462e0c67/tutorials/getting_started/laziness.md)

**Comment by schmmd** [comment]
Closing, since we recommend using a smaller dataset when developing.
